### PR TITLE
Add update revamp

### LINKED ISSUE #1487: Update "Add / Hide Tokens" screen
Zeplin:
https://app.zeplin.io/project/5d088205bff2d15de6a4397b/screen/5e209479cbb745a93fbda293/

This is how it looks now:

---

And this is how it should look like

Screenshot 2020-06-26 17 03 40

### PR DESCRIPTION
**Pull request by ChintanRathod**
Fixes #1487

### CODE REVIEW

**Comment by ChintanRathod** [comment]
Output

ezgif com-video-to-gif (2)

**Comment by JamesSmartCell** [comment]
What does the 'Jake Wharton' rx library do in this case? is it associated with the search feature?

File: app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java [hunk 1]
```diff
@@ -1,113 +1,339 @@
...
+public class TokenListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements Filterable {
+    private LayoutInflater inflater;
+    private Context context;
+    private ItemClickListener listener;
+    private final List<Token> tokens;
+    private ArrayList<Object> filteredData = new ArrayList<>();
+    private ArrayList<TokenSortedItem> displayTokensList = new ArrayList<>();
+    private ArrayList<TokenSortedItem> hiddenTokensList = new ArrayList<>();
+    protected final AssetDefinitionService assetService;
+
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java:52]: Deduplicate.

File: app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java [hunk 2]
```diff
@@ -1,113 +1,339 @@
...
+                    }else{
+                        hiddenTokensList.add(new TokenSortedItem(TokenViewHolder.VIEW_TYPE, t, t.getNameWeight()));
+                    }
                 }
             }
         }
 
-        Collections.sort(filteredList);
-        return filteredList;
+        if(performSorting) {
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java:81]: inconsistent formatting

File: app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java [hunk 3]
```diff
@@ -1,113 +1,339 @@
...
+        {
+            TokenSortedItem token = (TokenSortedItem) filteredData.get(position);
+            TokenViewHolder tokenViewHolder = (TokenViewHolder) viewHolder;
+            viewHolder.setIsRecyclable(false);
+            tokenViewHolder.tokenName.setText(token.value.getFullName());
+            tokenViewHolder.switchEnabled.setChecked(token.value.tokenInfo.isEnabled);
+            tokenViewHolder.switchEnabled.setTag(new Integer(position));
+            tokenViewHolder.switchEnabled.setOnCheckedChangeListener((v, b) -> {
+                listener.onItemClick(((TokenSortedItem) filteredData.get(position)).value, b);
+
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java:153]: formatting

File: app/src/main/java/com/alphawallet/app/ui/widget/holder/TokenHolder.java [hunk 4]
```diff
@@ -260,24 +255,15 @@ private void displayTokenIcon()
...
-            String tURL = assetDefinition.getTokenImageUrl(token.tokenInfo.chainId, token.getAddress());
-            if (TextUtils.isEmpty(tURL))
-            {
-                tURL = TRUST_ICON_REPO.replace(ICON_REPO_ADDRESS_TOKEN, correctedAddr);
-            }
-
-            boolean onlyTryCache = iconCheck.containsKey(correctedAddr);
+            IconItem iconItem = assetDefinition.fetchIconForToken(token);
 
             Glide.with(getContext().getApplicationContext())
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/ui/widget/holder/TokenHolder.java:260]: If we are going to use a new class for the iconItem then how about abstracting the code in .signature(...) out into the new class?

File: app/src/main/java/com/alphawallet/app/ui/TokenManagementActivity.java [hunk 5]
```diff
@@ -64,15 +91,42 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         hideZeroBalanceCheckBox.setOnCheckedChangeListener((v, checked) -> {
             pref.edit().putBoolean("hide_zero_balance_tokens", checked).apply();
         });
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/ui/TokenManagementActivity.java:93]: what does this do? Can you add a comment why we need the jake wharton library here?

Can we achieve the same thing without it using Runnable?

**Comment by JamesSmartCell** [comment]
@ChintanRathod can you write why we need to add the new 'jake wharton' library? Not sure what it's doing - can we achieve the same using runnables; making the code more maintainable?

File: app/src/main/java/com/alphawallet/app/ui/TokenManagementActivity.java (hunk 5 above)
Comment (ChintanRathod) [reply app/src/main/java/com/alphawallet/app/ui/TokenManagementActivity.java:93]: The RX Binding library binds Search EditText directly and apply filterable to it. Hence we used it inside the code. Definitely we can remove the overburden of library and implement ourselves.

File: app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java (hunk 1 above)
Comment (ChintanRathod) [reply app/src/main/java/com/alphawallet/app/ui/widget/adapter/TokenListAdapter.java:52]: I will make it work with single list

**Comment by ChintanRathod** [comment]
@JamesSmartCell I did necessary changes as mentioned except the IconItem one. I didn't get what is expected.

**Comment by JamesSmartCell** [comment]
@ChintanRathod One last thing can you update the 'confirm' button on this page to use FunctionButtonBar as per BackupKeyActivity etc.

**Comment by JamesSmartCell** [comment]
Hi @ChintanRathod I think this can be done in a much simpler way which is more expandable.  For example in this PR we are maintaining two lists for showing and hidden. When this view is expanded we'll need to add new lists for each, which is prone to failure and code bloat, and would need complex interchange mechanisms - having nightmares trying to work on it.

I took a subset of this PR, created a new branch to get only the icons working in the add/hide tokens view. This branch is merged into master.

How about using the views to achieve the lists. So for example you can have three weighting 'zones' each starting from Integer.MAXIMUM / 3, for the three elements we'll eventually have. Each title can have exactly the weighting of that start point, and entries under the list can follow on the base weighting.

Handled like this the code would be have a much smaller footprint and be more maintainable and easier to expand. 

Can you start again from the current master, which has a subset of this PR implemented; you'll need to re-do the titles entries you added before but this time the code should be a lot less complex. In theory it should be possible to implement only touching a couple of files.

Also - needs to use FunctionButtonBar.

**Pull request closed without merging**